christfellowshipchurch/apollos-app
# Group Customization

## DESCRIPTION

### 1. What does this PR do, or why is it needed?
- Adds an ⚙️ **Edit Group** button to a group details page for group leaders with access to the `GROUP_CUSTOMIZATION` feature flag
- Allows updating a Group's Cover Image from Rock-defined pool of options
- Allows adding and removing a Group's Resources (URL only for now)

### 2. What design trade-offs/decisions were made?

**Design**
The Edit Group button wasn't placed in a nuanced, ideal way. There is room for improvement, but for now, this was the least disruptive to the UX and technically simple.

I'd like to use the same round gear icon button from the bottom right corner of a users' avatar on Profile, but it was too tricky to get it a stable place for it.

**Tech**
Feel like there's a better way to [tackle the route components](https://github.com/christfellowshipchurch/apollos-app/pull/198/files#diff-c1c5b673d1406336829f0b1fee90e852bdc2c51b04f1e5ef6baa87451c28aa92R61) within `<EditGroup>` navigator.
Some redundancy and heavy-handedness.
Would also like to use a context provider to remove some prop drilling in Resources list/form components but not critical for now.

### 3. How do I test this PR?
1. Log in as a user who is, ideally, in multiple groups — one where they are a leader, and one where they aren't. User should also have access to `GROUP_CUSTOMIZATION` feature flag (pull latest from 🛰️ **api** `master` branch)
1. Go to a group page that you are **not** a leader of
1. 🔸 Verify there is no **Edit Group** button in the header area
1. Go back to a group you **are** the leader of
1. 🔸 Verify there's a new **Edit Group** button in the header area
1. Tap the button to edit the group
1. 🔸 Verify you can see a list of cover image options and the group's current photo, update it, cancel updating, etc
1. 🔸 Verify you can see a list of group resources, and you can remove and add new ones. Validate the form doesn't allow you to submit duplicate title/url combinations that already exist.

### 4. What automated tests did you write?
None

## SCREENSHOTS/GIFS

|Caption|Screenshot/GIF|
|---|---|
|Edit Cover Image flow|<img src="https://user-images.githubusercontent.com/1812955/99761114-46d0a400-2ac3-11eb-98b8-982a1d6548d2.gif" width="300" />|
|Group with Edit button|<img src="https://user-images.githubusercontent.com/1812955/99760953-ef323880-2ac2-11eb-9555-8e16fa952df9.jpg" width="300" />|
|Edit Group modal|<img src="https://user-images.githubusercontent.com/1812955/99760971-f9eccd80-2ac2-11eb-8a2a-a198c53e3101.jpg" width="300" />|
|Edit Group > Cover Image modal|<img src="https://user-images.githubusercontent.com/1812955/99760977-fce7be00-2ac2-11eb-9d9e-e8ab97d39bf1.jpg" width="300" />|
|Update cover image confirmation|<img src="https://user-images.githubusercontent.com/1812955/99760978-ffe2ae80-2ac2-11eb-9543-87d94e2531bc.jpg" width="300" />|
|Adding a new Resource|<img src="https://user-images.githubusercontent.com/1812955/99995047-e1f9a000-2d87-11eb-912b-32743b2bf81f.gif" width="300" />|
|Resources List and Add form screenshot|<img src="https://user-images.githubusercontent.com/1812955/99995156-fccc1480-2d87-11eb-9f11-5f99451a184b.png" width="300" />

## TODO:

- [ ] No new warnings in tests, storybook, or apps
- [ ] Assign relevant reviewers
- [ ] Upload GIF(s) of iOS, Android
- [ ] Get one peer approval before merging

## REVIEW:

**Manual QA**

- [ ] QA branch on iOS and ensure it looks/behaves as expected
- [ ] QA branch on Android and ensure it looks/behaves as expected

**Code Review**


> The purpose of PR Review is to _improve the quality of the software._
